File: src/lib/types/lastfm.ts
/**
 * Tipos TypeScript para la API de Last.fm
 */

import type { MusicFile } from '@/lib/types';

export interface LastFmImage {
  '#text': string;
  size: 'small' | 'medium' | 'large' | 'extralarge' | 'mega' | '';
}

export interface LastFmTag {
  name: string;
  url: string;
}

export interface LastFmArtistInfo {
  name: string;
  mbid?: string;
  url: string;
  image: LastFmImage[];
  stats?: {
    listeners: string;
    playcount: string;
  };
  bio?: {
    summary: string;
    content: string;
    published?: string;
  };
  tags?: {
    tag: LastFmTag[];
  };
}

export interface LastFmAlbumInfo {
  name: string;
  artist: string;
  mbid?: string;
  url: string;
  image: LastFmImage[];
  listeners?: string;
  playcount?: string;
  tags?: {
    tag: LastFmTag[];
  };
  wiki?: {
    summary: string;
    content: string;
    published?: string;
  };
  tracks?: {
    track: Array<{
      name: string;
      duration: string;
      url: string;
      streamable: { '#text': string; fulltrack: string };
      artist: {
        name: string;
        mbid: string;
        url: string;
      };
    }>;
  };
}

export interface LastFmTrackInfo {
  name: string;
  mbid?: string;
  url: string;
  duration?: string;
  listeners?: string;
  playcount?: string;
  artist: {
    name: string;
    mbid?: string;
    url: string;
  };
  album?: {
    artist: string;
    title: string;
    mbid?: string;
    url: string;
    image: LastFmImage[];
  };
  toptags?: {
    tag: LastFmTag[];
  };
  wiki?: {
    summary: string;
    content: string;
    published?: string;
  };
}

export interface LastFmError {
  error: number;
  message: string;
}

// Tipos procesados para usar en el UI
export interface ProcessedArtistInfo {
  name: string;
  image: string | null;
  bio: string;
  bioFull: string;
  tags: string[];
  listeners: number;
  playcount: number;
  url: string;
}

export interface ProcessedAlbumInfo {
  name: string;
  artist: string;
  image: string | null;
  summary: string;
  tags: string[];
  listeners: number;
  playcount: number;
  url: string;
  trackCount: number;
}

export interface ProcessedTrackInfo {
  name: string;
  artist: string;
  album: string | null;
  image: string | null;
  tags: string[];
  duration: number | null;
  listeners: number | null;
  playcount: number | null;
  wiki: string | null;
  url: string;
}

export interface EnrichedTrack {
  original: MusicFile;
  enriched: ProcessedTrackInfo | null;
  albumArtUrl: string | null;
}

File: src-tauri/src/domain/lastfm.rs
//! Last.fm domain types
#![allow(dead_code)]

use crate::domain::music::MusicFile;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProcessedTrackInfo {
    pub name: String,
    pub artist: String,
    pub album: Option<String>,
    pub duration: Option<u32>,
    pub playcount: Option<u64>,
    pub listeners: Option<u64>,
    pub tags: Vec<String>,
    pub wiki: Option<String>,
    pub url: String,
    pub image: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProcessedArtistInfo {
    pub name: String,
    pub image: Option<String>,
    pub bio: String,
    pub bio_full: String,
    pub tags: Vec<String>,
    pub listeners: u64,
    pub playcount: u64,
    pub url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProcessedAlbumInfo {
    pub name: String,
    pub artist: String,
    pub image: Option<String>,
    pub summary: String,
    pub tags: Vec<String>,
    pub listeners: u64,
    pub playcount: u64,
    pub url: String,
    pub track_count: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EnrichedTrack {
    pub original: MusicFile,
    pub enriched: Option<ProcessedTrackInfo>,
    pub album_art_url: Option<String>,
}

// Internal Last.fm API types for deserialization
#[derive(Debug, Deserialize)]
pub struct LastFmImage {
    #[serde(rename = "#text")]
    pub text: String,
    pub size: String,
}

#[derive(Debug, Deserialize)]
pub struct LastFmTag {
    pub name: String,
    pub url: String,
}

pub mod raw {
    use super::*;

    #[derive(Debug, Deserialize)]
    pub struct TrackResponse {
        pub track: TrackInfo,
    }

    #[derive(Debug, Deserialize)]
    pub struct TrackInfo {
        pub name: String,
        pub duration: Option<String>,
        pub listeners: Option<String>,
        pub playcount: Option<String>,
        pub artist: ArtistShort,
        pub album: Option<AlbumShort>,
        pub toptags: Option<TopTags>,
        pub wiki: Option<Wiki>,
        pub url: String,
    }

    #[derive(Debug, Deserialize)]
    pub struct ArtistResponse {
        pub artist: ArtistInfo,
    }

    #[derive(Debug, Deserialize)]
    pub struct ArtistInfo {
        pub name: String,
        pub image: Option<Vec<LastFmImage>>,
        pub bio: Option<Bio>,
        pub tags: Option<Tags>,
        pub stats: Option<Stats>,
        pub url: String,
    }

    #[derive(Debug, Deserialize)]
    pub struct AlbumResponse {
        pub album: AlbumInfo,
    }

    #[derive(Debug, Deserialize)]
    pub struct AlbumInfo {
        pub name: String,
        pub artist: String,
        pub image: Option<Vec<LastFmImage>>,
        pub wiki: Option<Wiki>,
        pub tags: Option<Tags>,
        pub listeners: Option<String>,
        pub playcount: Option<String>,
        pub url: String,
        pub tracks: Option<TracksWrapper>,
    }

    #[derive(Debug, Deserialize)]
    pub struct ArtistShort {
        pub name: String,
    }

    #[derive(Debug, Deserialize)]
    pub struct AlbumShort {
        pub title: String,
        pub image: Option<Vec<LastFmImage>>,
    }

    #[derive(Debug, Deserialize)]
    pub struct TopTags {
        pub tag: Vec<LastFmTag>,
    }

    #[derive(Debug, Deserialize)]
    pub struct Tags {
        pub tag: Vec<LastFmTag>,
    }

    #[derive(Debug, Deserialize)]
    pub struct Wiki {
        pub summary: String,
        pub content: String,
    }

    #[derive(Debug, Deserialize)]
    pub struct Bio {
        pub summary: String,
        pub content: String,
    }

    #[derive(Debug, Deserialize)]
    pub struct Stats {
        pub listeners: String,
        pub playcount: String,
    }

    #[derive(Debug, Deserialize)]
    pub struct TracksWrapper {
        pub track: Vec<TrackShort>,
    }

    #[derive(Debug, Deserialize)]
    pub struct TrackShort {
        pub name: String,
    }
}
